### ISSUE TITLE
DRAKE_DEMAND, DRAKE_THROW_UNLESS, etc don't play nicely with drake::optional

### ISSUE THREAD

**Issue by RussTedrake** [RobotLocomotion/drake#8170]
Their check for "should be bool-convertible" fails.  
Currently have to call, e.g.
```
DRAKE_DEMAND(static_cast<bool>(my_optional_variable))
```
instead of just 
```
DRAKE_DEMAND(my_optional_variable)
```

**Comment by jwnimmer-tri** [comment]
The `operator bool()` is by design marked as `explicit` (http://en.cppreference.com/w/cpp/utility/optional).

For clarity of the condition, I'd like to fix this via tcbrindle/cpp17_headers#13 and then `DRAKE_DEMAND(my_optional_variable.has_value());`.

For now, if we want to be terse, we can do `DRAKE_DEMAND(!!my_optional_variable);`.

**Comment by RussTedrake** [comment]
SGTM
